Selection handles position is off

chunhtai opened the pull request:
… non-obscureText

## Description

Added a logic to check whether an update to selection overlay is necessary whenever editable text gets rebuilt.

## Related Issues

https://github.com/flutter/flutter/issues/34607

## Tests

I added the following tests:
'selection overlay will update when text grow bigger'

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I signed the [CLA].
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [ ] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

breaking change draft: https://docs.google.com/document/d/1JI-5idfRxzqcue3euGeJBuCtsEhnX7_VMeeb6Qp2Bwg/edit

goderbauer commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -912,9 +912,10 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
       _updateRemoteEditingValueIfNeeded();
-    }
-    if (widget.controller.selection != oldWidget.controller.selection) {
+    } else if (widget.controller.selection != oldWidget.controller.selection) {
```
Why else if? Don't we need to call _selectionIOverlay.update if the controller and the selection on it changed? (we should add a test to verify that this works)

goderbauer commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -684,11 +710,13 @@ class _TextSelectionHandleOverlayState
         point = endpoints[0].point;
         type = _chooseType(endpoints[0], TextSelectionHandleType.left, TextSelectionHandleType.right);
+        // Only update endpoints in first handle that always gets built.
+        widget.onEndpointsUpdate(endpoints);
```
Piping this back like this feels strange...

goderbauer commented on packages/flutter/test/widgets/editable_text_test.dart:
```diff
@@ -268,6 +268,59 @@ void main() {
+    await tester.idle();
+
+    List<CompositedTransformFollower> container =
+    find.byType(CompositedTransformFollower)
```
nit: indentation

goderbauer replied:
tester.widgetList(find.byType(...))?

goderbauer commented on packages/flutter/test/widgets/editable_text_test.dart:
```diff
@@ -268,6 +268,59 @@ void main() {
+    await tester.pumpAndSettle();
+
+    // Handles should be updated with bigger font size.
+    container = find.byType(CompositedTransformFollower)
```
Same here.

goderbauer commented:
Could this be solved by having the RenderEditable insert LeaderLayers [1] where the selection handles are supposed to go and the selection handles itself are drawn on a FollowerLayer [2], which is always positioned correctly to be in the right place even when the text size in the RenderEditable changes?

See also the Widgets CompositedTransformTarget [3] and CompositedTransformFollower [4], which insert these kind of layers.

Ian mentioned that @xster was talking about doing this at some point to fix a similar class of problems.

[1] https://master-api.flutter.dev/flutter/rendering/LeaderLayer-class.html
[2] https://master-api.flutter.dev/flutter/rendering/LeaderLayer-class.html
[3] https://master-api.flutter.dev/flutter/widgets/CompositedTransformTarget-class.html
[4] https://master-api.flutter.dev/flutter/widgets/CompositedTransformFollower-class.html

chunhtai committed:
fix Selection handles position is off

xster commented:
+1. @matthew-carroll had a prototype implementing the iOS text selection magnifier using this approach.

chunhtai commented:
finished refactoring ready for review

justinmc commented:
This made the Cupertino handles appear on the next line when I tried it:

<img width="200" src="https://user-images.githubusercontent.com/389558/60120398-8b70d800-9735-11e9-8bfb-59ae0f3d9502.png" />

chunhtai commented:
> This made the Cupertino handles appear on the next line when I tried it:
> 
> <img alt="" width="200" src="https://user-images.githubusercontent.com/389558/60120398-8b70d800-9735-11e9-8bfb-59ae0f3d9502.png">

thanks for catching this bug, I refined the offset calculation.

justinmc commented:
Thanks for the fix!  The handles look right to me now.

justinmc commented:
Another thing I'm noticing is that the interactivity improvements that I made in https://github.com/flutter/flutter/pull/31852 seem to be gone here.  The handles are very hard to touch for both Material and Cupertino.

chunhtai commented:
> Another thing I'm noticing is that the interactivity improvements that I made in #31852 seem to be gone here. The handles are very hard to touch for both Material and Cupertino.

It turns out I removed a bunch of them when i was experimenting the offset calculation. Added back

justinmc commented:
Thanks, the handles feel great now.

I've got one more "QA" bug for you, sorry for the bombardment haha.  The handle sometimes appears outside the input for both TextField and CupertinoTextField:

![handle_outside](https://user-images.githubusercontent.com/389558/60281936-b2a4e200-98ba-11e9-81c3-c9d1db438cb3.gif)

justinmc commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -143,6 +144,7 @@ class RenderEditable extends RenderBox {
     Color backgroundCursorColor,
     ValueNotifier<bool> showCursor,
     bool hasFocus,
+    @required List<LayerLink> handleLayerLinks,
```
Is this a breaking change?

chunhtai replied:
yes

justinmc commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -269,8 +269,10 @@ class TextSelectionOverlay {
     this.debugRequiredFor,
-    @required this.layerLink,
+    @required this.toolbarLayerLink,
+    @required this.handleLayerLinks,
```
More potential breaking changes.

justinmc reviewed (commented):
See my concern about breaking changes.  Otherwise this looks pretty good and seems to work great now.  I like the LayerLink approach.

goderbauer commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -1765,13 +1779,38 @@ class RenderEditable extends RenderBox {
+    context.pushLayer(
+        LeaderLayer(link: handleLayerLinks[0], offset: startPoint),
+        super.paint,
+        Offset.zero
```
nit: add trailing comma and indent by two spaces.

goderbauer commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -1765,13 +1779,38 @@ class RenderEditable extends RenderBox {
+      context.pushLayer(
+          LeaderLayer(link: handleLayerLinks[1], offset: endPoint),
+          super.paint,
+          Offset.zero
```
same.

goderbauer commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -143,6 +144,7 @@ class RenderEditable extends RenderBox {
     Color backgroundCursorColor,
     ValueNotifier<bool> showCursor,
     bool hasFocus,
+    @required List<LayerLink> handleLayerLinks,
```
Since we always require two: would it be more meaningful to just have to separate parameters with meaningful names instead of a list?

goderbauer replied:
(e.g. leading/trailingHandleLayerLink)?

goderbauer replied:
same elsewhere where the list is used.

goderbauer commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -1224,9 +1227,11 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
         debugRequiredFor: widget,
-        layerLink: _layerLink,
+        toolbarLayerLink: _toolbarLayerLink,
+        handleLayerLinks : _handleLayerLinks,
```
nit: remove space before :

goderbauer commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -300,13 +303,21 @@ class TextSelectionOverlay {
   /// The editable line in which the selected text is being displayed.
   final RenderEditable renderObject;
 
+  /// The object supplied to [_TextSelectionHandleOverlay] that determines the
```
This is surprising, why does the textDirection determine the handle type?

chunhtai replied:
textdirection will change the order of handles
ltr:
start -> left handle,  end -> right handle
rtl:
start -> right handle, end -> left handle

chunhtai replied:
but I think i can avoid this parameter, I was thinking of getting rid of renderObject dependency from this class, but it seems I will still need it to calculate the handlesize

goderbauer reviewed (commented):
Please update the PR with the breaking changes.

chunhtai commented:
Addressed comment, attached break change email draft. Ready for review

justinmc reviewed (approved):
LGTM.  I requested access to the breaking change email, so I'll take a look at that later.

Update: Got access and it looks good too!

goderbauer commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -903,6 +910,32 @@ class RenderEditable extends RenderBox {
     markNeedsPaint();
   }
 
+  /// The layer link of start selection handle.
```
nit `[LayerLink]`

goderbauer commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -903,6 +910,32 @@ class RenderEditable extends RenderBox {
+    markNeedsPaint();
+  }
+
+  /// The layer link of end selection handle.
```
same

goderbauer commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -848,7 +848,12 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
 
-  final LayerLink _layerLink = LayerLink();
+  final LayerLink _toolbarLayerLink = LayerLink();
+
```
nit: Just group these three together without blank lines in between since they are logically similar.

goderbauer reviewed (approved):
LGTM
